ucp/_libs/ucx_api.pyx

```diff
@@ -0,0 +1,475 @@
+    if req.data == NULL:
+        # This callback function was called before handle_comm_result
+        # had a chance to set req.data
+        req.finished = True
+        req.received = length
+        return
+    cdef object req_data = <object> req.data
+
+    log_str = None
+    msg = "Error receiving%s " %(" \"%s\":" % log_str if log_str else ":")
+    if status == UCS_ERR_CANCELED:
+        exception = UCXCanceled()
+    elif status != UCS_OK:
+        msg += ucs_status_string(status).decode("utf-8")
+        exception = UCXError(msg)
+    else:
+        exception = None
+
+    ucp_request_reset(request)
+    ucp_request_free(request)
+    size = length if exception is None else 0
+    req_data["cb_func"](exception, size, *req_data["cb_args"])
+    Py_DECREF(req_data)
+
+
+cdef void _ucx_tag_recv_callback(void *request, ucs_status_t status,
+                                 ucp_tag_recv_info_t *info):
+    _ucx_recv_callback(request, status, info.length)
+
+
+cdef uintptr_t handle_comm_result(
+    ucs_status_ptr_t status, dict req_data, expected_receive=None
+):
+    log_str = None
+    msg = "Comm Error%s " %(" \"%s\":" % log_str if log_str else ":")
+    exception = None
+    if UCS_PTR_STATUS(status) == UCS_OK:
+        pass
+    elif UCS_PTR_IS_ERR(status):
+        msg += ucs_status_string(UCS_PTR_STATUS(status)).decode("utf-8")
+        exception = UCXError(msg)
+    else:
+        req = <ucp_request*> status
+        if req.finished:  # The callback function has already handle the request
+            ucp_request_reset(req)
+            ucp_request_free(req)
+        else:
+            # The callback function has not been called yet.
+            # We fill `ucp_request` for the callback function to use
+            Py_INCREF(req_data)
+            req.data = <PyObject*> req_data
+            return <uintptr_t><void*>req
+
+    if expected_receive:
+        req_data["cb_func"](exception, expected_receive, *req_data["cb_args"])
+    else:
+        req_data["cb_func"](exception, *req_data["cb_args"])
+    return 0
+
+
+cdef void _ucx_send_callback(void *request, ucs_status_t status):
+    cdef ucp_request *req = <ucp_request*> request
+    if req.data == NULL:
+        # This callback function was called before handle_comm_result
+        # had a chance to set req.data
+        req.finished = True
+        return
+
+    cdef object req_data = <object> req.data
+
+    if status == UCS_ERR_CANCELED:
+        exception = UCXCanceled()
+    elif status != UCS_OK:
+        log_str = None
+        msg = "Error sending%s " %(" \"%s\":" % log_str if log_str else ":")
+        msg += ucs_status_string(status).decode("utf-8")
+        exception = UCXError(msg)
+    else:
+        exception = None
+
+    ucp_request_reset(request)
+    ucp_request_free(request)
+    req_data["cb_func"](exception, *req_data["cb_args"])
+    Py_DECREF(req_data)
+
+
+def ucx_tag_send(UCXEndpoint ep, buffer, size_t nbytes,
```

Comment:
I see `ucx_stream_{send,recv}`, but only `ucx_tag_send`, is `ucx_tag_recv` missing or is this intended?

Reply:
`ucx_tag_recv()` was a method instead `UCXWorker`, moved it out to the other functions